Enable ufw and update allowed/denied port rules

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR does the following things:
* Enables `ufw` (previously, we left `ufw` disabled)
* Modifies the list of ports allowed/denied by `ufw` to those specified in https://github.com/cisagov/cool-system-internal/issues/80

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

These changes were requested by our Assessments team in order to make these instances more secure when they are initially deployed.  This work is part of https://github.com/cisagov/cool-system-internal/issues/80.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I deployed a Teamserver instance with these changes to a Staging environment and confirmed that `ufw` was active when the instance started up and the expected rules were set:

```
$ sudo ufw status numbered
Status: active

     To                         Action      From
     --                         ------      ----
[ 1] 22/tcp                     ALLOW IN    Anywhere                  
[ 2] 25/tcp                     ALLOW IN    Anywhere                  
[ 3] 587/tcp                    ALLOW IN    Anywhere                  
[ 4] 993/tcp                    ALLOW IN    Anywhere                  
[ 5] 5000:5999/tcp              ALLOW IN    Anywhere                  
[ 6] 5901/tcp                   ALLOW IN    Anywhere                  
[ 7] 50050/tcp                  ALLOW IN    Anywhere                  
[ 8] 53/tcp                     DENY IN     Anywhere                  
[ 9] 80/tcp                     DENY IN     Anywhere                  
[10] 443/tcp                    DENY IN     Anywhere                  
[11] 8080/tcp                   DENY IN     Anywhere                  
[12] 53/udp                     DENY IN     Anywhere                  
[13] 8080/udp                   DENY IN     Anywhere                  
[14] 22/tcp (v6)                ALLOW IN    Anywhere (v6)             
[15] 25/tcp (v6)                ALLOW IN    Anywhere (v6)             
[16] 587/tcp (v6)               ALLOW IN    Anywhere (v6)             
[17] 993/tcp (v6)               ALLOW IN    Anywhere (v6)             
[18] 5000:5999/tcp (v6)         ALLOW IN    Anywhere (v6)             
[19] 5901/tcp (v6)              ALLOW IN    Anywhere (v6)             
[20] 50050/tcp (v6)             ALLOW IN    Anywhere (v6)             
[21] 53/tcp (v6)                DENY IN     Anywhere (v6)             
[22] 80/tcp (v6)                DENY IN     Anywhere (v6)             
[23] 443/tcp (v6)               DENY IN     Anywhere (v6)             
[24] 8080/tcp (v6)              DENY IN     Anywhere (v6)             
[25] 53/udp (v6)                DENY IN     Anywhere (v6)             
[26] 8080/udp (v6)              DENY IN     Anywhere (v6)
```

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been approved. -->

- [x] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Add a tag or create a release.

dav3r committed:
Deny any ports not needed for Guacamole and mail to function properly

Also, allow some ports that may be used internally by assessors.

dav3r committed:
Enable ufw and enforce firewall rules by default

It's preferable to have the default state of these instances be locked down with ufw enabled.

dav3r committed:
Tweak wording for consistency

dav3r committed:
Bump version from 0.2.8 to 0.3.0

dav3r committed:
Bump version from 0.3.0 to 0.3.0-rc.1

jsf9k commented on src/ufw.yml:
```diff
@@ -6,36 +6,46 @@
-    # Do not actually enforce firewall rules.  The assessors will turn
-    # UFW on if they want it.
-    state: disabled
+    # Enable ufw and enforce firewall rules by default.
+    state: enabled
```
These lines can be removed.  `enabled` is [the default behavior](https://github.com/cisagov/ansible-role-ufw/blob/41e0944c9f2d17ce58e68812f1cc0386b6d32872/defaults/main.yml#L11-L13).

dav3r replied:
Cleaned up in https://github.com/cisagov/teamserver-packer/pull/52/commits/52db4be9c4c96d0fd9851f6e94ecd30831bb74bf.

jsf9k commented on src/ufw.yml:
```diff
@@ -6,36 +6,46 @@
+        - "80"       # Domain verification (external)
+        - "443"      # Payload hosting (external)
         - "8080"     # C2 listening and testing (external)
         - 8000:8999  # Additional listeners, if necessary (external)
```
My understanding was that they are using 5000-5999 now _instead of_ 8000-8999, in which case this line can be removed.

dav3r replied:
@m1j09830 Can you please confirm if you still want 8000-8999 on this list of denied ports?  I know you wrote that you did in https://github.com/cisagov/cool-system-internal/issues/80, but I just want to double-check since @jsf9k brings up a good point here.

m1j09830 replied:
Hello all, for the TeamServer you are correct, however since NOM reversed the change that was requested a while back our TeamServers currently still have port range 8000-8999 exposed to the world. So in an abundance of caution I've included that be part of the deny list until it's formally removed from the build process.

jsf9k replied:
@dav3r - If the 8000-8999 port range stays then please add a comment explaining [what @m1j09830 stated above](https://github.com/cisagov/teamserver-packer/pull/52#discussion_r796924230).

m1j09830 replied:
@jsf9k after speaking with @dav3r about it I went ahead and removed that range from the initial ticket. I'm going to coordinate with NOM to implement our original request of removing ingress over that port range now that you all have implemented the internal range for us to use. I've updated the original ticket to reflect the change.

dav3r replied:
I spoke with @m1j09830 and he agreed that we will no longer need to include 8000-8999 here, so I will remove it.  He has already updated [his comment in cisagov/cool-system-internal#80](https://github.com/cisagov/cool-system-internal/issues/80#issuecomment-1025951812).

dav3r replied:
I removed that port range in https://github.com/cisagov/teamserver-packer/pull/52/commits/4d0cf1401da251c048cf040cee7e1aba73dd5d0f.

jsf9k commented on src/ufw.yml:
```diff
@@ -6,36 +6,46 @@
+      # These rules may be manually modified by assessors as needed, but we
+      # want to default to a minimal set of allowed ports.
+      community.general.ufw:
+        proto: tcp
+        rule: deny
+        to_port: "{{ item }}"
+      loop:
+        - "53"       # C2 traffic (external)
+        - "80"       # Domain verification (external)
+        - "443"      # Payload hosting (external)
         - "8080"     # C2 listening and testing (external)
         - 8000:8999  # Additional listeners, if necessary (external)
-        - "50050"    # CobaltStrike (internal to assessment env)
-    - name: Allow incoming traffic through default UDP ports
+    - name: Deny incoming traffic to specified UDP ports
+      # These rules may be manually modified by assessors as needed, but we
+      # want to default to a minimal set of allowed ports.
       community.general.ufw:
         proto: udp
-        rule: allow
+        rule: deny
         to_port: "{{ item }}"
       loop:
         - "53"    # C2 traffic (external)
```
This section can be removed as well.  The default behavior for incoming traffic is `deny`.

jsf9k replied:
I assume this is due to the same jazz as [here](https://github.com/cisagov/kali-packer/pull/92#discussion_r796747165).  Please add a comment, just as I requested [here](https://github.com/cisagov/kali-packer/pull/92#discussion_r796798197).

dav3r replied:
Comment added in https://github.com/cisagov/teamserver-packer/pull/52/commits/16f2c947f9cb468a167a82a48892628484731c2d.

jsf9k reviewed (commented):
Some comments for your consideration.

dav3r committed:
Remove ufw state variable 

We are no longer changing it from its default value, so it is no longer needed.

Co-Authored-By: Shane Frasier <maverick@maverickdolphin.com>

dav3r committed:
Add a comment explaining why we create "deny" ufw rules

Co-Authored-By: Shane Frasier <maverick@maverickdolphin.com>

dav3r committed:
Remove TCP ports 8000-8999 from the set of deny rules

Co-Authored-By: Shane Frasier <maverick@maverickdolphin.com>

dav3r committed:
Bump version from 0.3.0-rc.1 to 0.3.0-rc.2

jsf9k reviewed (approved):
Great work!

mcdonnnj commented on src/ufw.yml:
```diff
@@ -5,37 +5,47 @@
+        rule: deny
+        to_port: "{{ item }}"
+      loop:
+        - "53"       # C2 traffic (external)
+        - "80"       # Domain verification (external)
+        - "443"      # Payload hosting (external)
+        - "8080"     # C2 listening and testing (external)
```
Since this is a `deny` list I don't think the `(external)` markers make sense. I also thought the entry descriptions should mirror the linked issue. 

```suggestion
        - "53"       # DNS
        - "80"       # Domain verification
        - "443"      # Payload hosting
        - "8080"     # C2 listening
```

mcdonnnj reviewed (approved):
Makes sense. I had one commenting suggestion for consideration.

dav3r committed:
Improve comments

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Bump version from 0.3.0-rc.2 to 0.3.0

dav3r committed:
Merge branch 'improvement/enable-ufw' of github.com:cisagov/teamserver-packer into improvement/enable-ufw

jsf9k commented:
Please start to prefer rebases to merge commits.  Merge commits are the devil!
